File: backend/app/utils/validators.py
import re
import uuid
from typing import Any, Optional, List, Dict, Union
from datetime import datetime
from urllib.parse import urlparse

def validate_username(username:str) -> bool:
    if not username:
        return False
    if len(username) < 3 or len(username) > 50:
        return False
    
    pattern = r'^[a-zA-Z][a-zA-Z0-9_]*$'
    return bool(re.match(pattern,username))

def validate_password(password:str) -> Dict[str,Any]:
    return{
        "min_length": len(password) >= 8,
        "has_uppercase": any(c.isupper() for c in password),
        "has_lowercase": any(c.islower() for c in password),
        "has_digit": any(c.isdigit() for c in password),
        "has_special": any(c in '!@#$%^&*()_+-=[]{};:,.<>?' for c in password),
        "no_common": password.lower() not in ['password', '12345678', 'qwerty123', 'admin123'],
    }
def is_password_strong(password: str) -> bool:
    validation = validate_password(password)
    return all(validation.values())

def validate_url(url: str) -> bool:
    if not url:
        return False
    
    try:
        result = urlparse(url)
        return all([result.scheme, result.netloc])
    except:
        return False

def validate_image_file(filename: str) -> bool:
    allowed_extensions = {'.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp', '.tiff'}
    return any(filename.lower().endswith(ext) for ext in allowed_extensions)

def validate_video_file(filename: str) -> bool:
    allowed_extensions = {'.mp4', '.mov', '.avi', '.webm', '.mkv', '.flv', '.mpeg'}
    return any(filename.lower().endswith(ext) for ext in allowed_extensions)

def validate_file_size(file_size: int, max_size_mb: int = 50) -> bool:
    return file_size <= max_size_mb * 1024 * 1024

def validate_media_type(content_type: str) -> str:
    if content_type.startswith('image/'):
        return 'image'
    elif content_type.startswith('video/'):
        return 'video'
    elif content_type.startswith('audio/'):
        return 'audio'
    else:
        return 'unknown'

def validate_date_range(start_date: datetime, end_date: datetime) -> bool:
    return start_date < end_date

def validate_time_range_days(days: int, min_days: int = 1, max_days: int = 180) -> bool:
    return min_days <= days <= max_days

def is_valid_iso_date(date_str: str) -> bool:
    try:
        datetime.fromisoformat(date_str)
        return True
    except:
        return False

def validate_uuid(uuid_str: str) -> bool:
    try:
        uuid.UUID(uuid_str)
        return True
    except ValueError:
        return False

def validate_uuid_list(uuid_list: List[str]) -> List[bool]:
    return [validate_uuid(u) for u in uuid_list]

def validate_json(data: Any) -> bool:
    try:
        import json
        json.dumps(data)
        return True
    except:
        return False

def validate_platform(platform: str) -> bool:
    valid_platforms = {
        'instagram', 'facebook', 'twitter', 'tiktok', 
        'telegram', 'reddit', 'youtube', 'linkedin'
    }
    return platform.lower() in valid_platforms

def validate_platforms(platforms: List[str]) -> List[str]:
    return [p for p in platforms if validate_platform(p)]

def sanitize_input(text: str, max_length: int = 1000) -> str:
    if not text:
        return ""
    
    text = text.strip()
    
    if len(text) > max_length:
        text = text[:max_length]
    
    dangerous = ['<', '>', '&', '"', "'", '`']
    for char in dangerous:
        text = text.replace(char, '')
    
    return text

def sanitize_filename(filename: str) -> str:
    filename = filename.replace('/', '').replace('\\', '')
    
    import re
    filename = re.sub(r'[^\w\-_.]', '', filename)
    
    return filename

def sanitize_email(email: str) -> str:
    if not email:
        return ""
    
    email = email.strip().lower()
    return email

__all__ = [
    "validate_email",
    "validate_username",
    "validate_password",
    "is_password_strong",
    
    "validate_url",
    "validate_image_file",
    "validate_video_file",
    "validate_file_size",
    "validate_media_type",
    
    "validate_date_range",
    "validate_time_range_days",
    "is_valid_iso_date",
    
    "validate_uuid",
    "validate_uuid_list",
    
    "validate_json",
    "validate_platform",
    "validate_platforms",
    
    "sanitize_input",
    "sanitize_filename",
    "sanitize_email",
]

File: client/src/utils/validators.ts
export const validateEmail = (email: string): boolean => {
    if (!email) return false;
    const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
    return pattern.test(email);
};

export const validateUsername = (username: string): boolean => {
    if (!username) return false;
    if (username.length < 3 || username.length > 50) return false;
    const pattern = /^[a-zA-Z0-9_]+$/;
    return pattern.test(username);
};

export const validatePassword = (password: string) => {
    return {
        minLength: password.length >= 8,
        maxLength: password.length <= 128,
        hasUpper: /[A-Z]/.test(password),
        hasLower: /[a-z]/.test(password),
        hasNumber: /[0-9]/.test(password),
        hasSpecial: /[!@#$%^&*()_+\-=\[\]{};:,.<>?]/.test(password),
        noCommon: !['password', '12345678', 'qwerty123', 'admin123'].includes(
            password.toLowerCase()
        ),
    };
};

export const isPasswordStrong = (password: string): boolean => {
    if (!password) return false;
    const validate = validatePassword(password)
    return Object.values(validate).every((value) => value === true);
};

export const validateUrl = (url: string): boolean => {
    if (!url) return false;
    try {
        const parsed = new URL(url);
        return ["http:", "https:", "ftp:", "mailto:"].includes(parsed.protocol);
    } catch (e) {
        return false;
    }
};

export const validateImageFile = (filename: string): boolean => {
    const allowed = ['.jpg', '.jpeg', '.png', '.gif', '.webp', '.avif'];
    const ext = filename.slice(filename.lastIndexOf('.')).toLowerCase();
    return allowed.includes(ext);
};

export const validateVideoFile = (filename: string): boolean => {
    const allowed = ['.mp4', '.mov', '.avi', '.webm', '.mkv', '.flv', '.mpeg', '.wmv'];
    const ext = filename.slice(filename.lastIndexOf('.')).toLowerCase();
    return allowed.includes(ext);
};

export const validateMediaFile = (contentType: string): 'image' | 'video' | 'audio' | 'document' | 'unknown' => {
    if (!contentType) return 'unknown';

    const prefix = contentType.split('/')[0].toLowerCase();
    const mimeType = contentType.split('/')[1].toLowerCase();

    if (prefix === 'image') return 'image';
    if (prefix === 'video') return 'video';
    if (prefix === 'audio') return 'audio';

    const docPatterns = ['pdf', 'doc', 'docx', 'txt', 'xls', 'xlsx', 'csv', 'md', 'rtf'];
    if (docPatterns.includes(mimeType)) return 'document';

    return 'unknown';
};

export const validateFileCount = (files: File[], maxCount = 1): boolean => {
    if (!files || files.length === 0) return false;
    if (maxCount === 0) return true;
    return files.length <= maxCount;
};

export const validateDateRange = (startDate: Date, endDate: Date): boolean => {
    return startDate < endDate;
};

export const validateTimeRangeDays = (days: number, minDays = 1, maxDays = 180): boolean => {
    return Number.isInteger(days) && days >= minDays && days <= maxDays;
};

export const isValidISODate = (dateStr: string): boolean => {
    try {
        const date = new Date(dateStr);
        return !isNaN(date.getTime());
    } catch {
        return false;
    }
};


export const validateUUID = (uuid: string): boolean => {
    const pattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
    return pattern.test(uuid);
};

export const validateUUIDList = (uuids: string[]): boolean[] => {
    return uuids.map(validateUUID);
};

export const validateJson = (data: any): boolean => {
    try {
        JSON.stringify(data);
        return true;
    } catch {
        return false;
    }
};

export const validatePlatform = (platform: string): boolean => {
    const valid = [
        'instagram', 'facebook', 'twitter', 'tiktok',
        'telegram', 'reddit', 'youtube', 'snapchat',
        'linkedin', 'pinterest', 'discord', 'whatsapp',
    ];
    return valid.includes(platform.toLowerCase());
};

export const validatePlatforms = (platforms: string[]): string[] => {
    return platforms.filter(validatePlatform);
};

export const validateConfidence = (confidence: number): boolean => {
    return confidence >= 0 && confidence <= 1;
};

export const sanitizeInput = (text: string, maxLength = 1000): string => {
    if (!text) return '';
    let sanitized = text.trim();
    if (sanitized.length > maxLength) {
        sanitized = sanitized.slice(0, maxLength);
    }
    return sanitized;
};

export const sanitizeFilename = (filename: string): string => {
    if (!filename) return '';
    return filename
        .replace(/[\/\\]/g, '')
        .replace(/[^\w\-_.]/g, '')
        .slice(0, 255);
};

export const sanitizeEmail = (email: string): string => {
    if (!email) return '';
    return email.trim().toLowerCase();
};

export const sanitizeHtml = (html: string): string => {
    if (!html) return '';
    return html
        .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
        .replace(/<[^>]+>/g, '')
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#x27;');
};